### PR TITLE
Enhancement: improve auto compatibility with self-caching hosts & add compatibility with wp.com

### LINKED ISSUE #2814: Auto-compatibillity with WordPress.com hosting
**Describe the solution you'd like**
WordPress.com is using their own page caching system. Because of that, WP Rocket is banned from this hosting. When someone activate WP Rocket, a notice is displayed to let the user know the plugin has been automatically disable to avoid conflict with their caching system.

To avoid any conflicts, and be possible to activate WP Rocket on wp.com, we must disable our page caching if the current website is hosted on wp.com 

**Additional context**
If possible, we should not try to write anything into advanced-cache.php. Having an empty advanced-cache.php like we are doing on WP Engine isn't a solution. This file is used by their page caching system.

On wp.com, another performance plugin is automatically installed on new website: [Page Optimize](https://wordpress.org/plugins/page-optimize/). This plugin can concatenate CSS and JS + add `async` or `defer` on JS files.

We will have to grey out our options if we notice these options are activated (like we do with Autoptimize).

### LINKED ISSUE #2855: Prevent creation of advanced-cache.php in WP Engine
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version 👍 
- Used the search feature to ensure that the bug hasn’t been reported before 👍 

**Describe the bug**
- On WP Engine, we do not need to write anything into `advanced-cache.php`, but still we create the file on WP Rocket activation, deactivation and settings update.
- With WP Rocket 3.6.1, we actually write contents into the file. (Issue with some more details here: https://github.com/wp-media/wp-rocket/issues/2896)

Video: https://youtu.be/6TqcUQS1M6k

**To Reproduce**
- Install WP Rocket 3.4.4 on WP Engine. Empty `advanced-cache.php` is created on activation and deactivation
- Install WP Rocket 3.6.1 on WP Engine. `advanced-cache.php` is not empty. 

**Expected behavior**
Since we do not intend to use the file on WP Engine as per: https://github.com/wp-media/wp-rocket/blob/7838b137afad05907f6281d114c72f8c34423de8/inc/ThirdParty/Hostings/WPEngine.php#L39

we should not create the file, nor modify it. 

**Additional context**
Feedback from Remy - https://wp-media.slack.com/archives/GUT7FLHF1/p1593444250001500

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by iCaspar**
This PR addresses issues #2814 and #2299 and #2855 

- Adds a filter `rocket_generate_advanced_cache_file` that can be set to `false` in order to short-circuit writing the `advanced-cache.php` file.
- Adds a `WordPressCom` 3rd-party Subscriber to handle wordpress.com-specific compatibility

With the new filter, we are also able to address issues on any hosts (eg. WordPress.com, WPEngine) where other caching is enabled, for which we should not be writing the `advanced-cache.php` file.

At the same time, part of the code was refactored to use a common base for all hosts with their own cache (see new abstract class `ThirdParty\Hostings\AbstractNoCacheHost`)

### Acceptance criteria
on Pressable, wp.com, WPEngine:
- Nothing written in both `wp-config.php` & `advanced-cache.php` at any time (activation/deactivation/WPR settings update)
- No admin notices about `wp-config.php` & `advanced-cache.php` writing permissions
- Server cache should be cleared when WP Rocket clear cache is triggered
- Varnish add-on section is not displayed
- WP Rocket cache files are not created

### CODE REVIEW

**Comment by iCaspar** [comment]
@Tabrisrp Note that in the new filter, I've set the default to `true` and the short-circuit to `false` -- it seemed more intuitive to say `true` => `write the file` and `false` => `don't write the file`.

File: inc/main.php [hunk 1]
```diff
@@ -221,10 +225,17 @@ function rocket_activation() {
 	rocket_init_config_dir();
 
 	// Create advanced-cache.php file.
+	if ( rocket_get_constant( 'WPCOMSH_VERSION' ) ) {
+		/**
+		 * This filter is documented in inc/files.php::rocket_generate_advanced_cache_file
+		 */
+		add_filter( 'rocket_generate_advanced_cache_file', __return_false() );
```
Comment (Tabrisrp) [inline inc/main.php:232]:
```suggestion
		add_filter( 'rocket_generate_advanced_cache_file', '__return_false' );
```

File: inc/ThirdParty/Hostings/WordPressCom.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	 * @return array The array of subscribed events.
+	 */
+	public static function get_subscribed_events() {
+		if ( ! rocket_get_constant( 'WPCOMSH_VERSION' ) ) {
+			return [];
+		}
+
+		return [
+			'do_rocket_generate_caching_files'    => 'return_false',
+			'rocket_cache_mandatory_cookies'      => 'return_false',
```
Comment (Tabrisrp) [inline inc/ThirdParty/Hostings/WordPressCom.php:68]:
```suggestion
			'rocket_cache_mandatory_cookies'      => 'return_empty_array',
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Comment by iCaspar** [comment]
Adding `Blocked` until we complete the new work on the Activate/Deactivate refactoring going on in PR #2860

**Comment by GeekPress** [comment]
@iCaspar @Tabrisrp  A message shared by Lucy:

> Just in case..... their support said we should be able to use this: https://developer.wordpress.org/reference/functions/wp_cache_flush/

**Comment by Tabrisrp** [comment]
One more thing to do in this PR, after [this comment](https://github.com/wp-media/wp-rocket/pull/2860#pullrequestreview-454431642)

For `Pressable`:
- Add the `Activation/DeactivationInterface`, and implement the 2 methods
- As we do for `WPEngine`, prevent writing the `WP_CACHE`constant in wp-config.php: https://github.com/wp-media/wp-rocket/blob/1afc7ebdf3b5166b74c9a7f8f172555a7cbb5aae/inc/ThirdParty/Hostings/WPEngine.php#L70

In `Engine\Cache\WPCache::should_set_wp_cache_constant()`, remove the check for `IS_PRESSABLE`, as it will be directly handled by the `Pressable` class

In `Activation` & `Deactivation` classes, add `Pressable` in the static `$(de)activators` array, and add the class to the service providers too

**Review by remyperona — APPROVED** [review]

**Comment by Tabrisrp** [comment]
@wp-media/qa Ready for QA, acceptance criteria are in the original post

**Review by Davidacu — COMMENTED** [review]
The Test execution details can be found [here](https://docs.google.com/spreadsheets/d/1kkoOx8tKpVurH7h9B50-AN9JdHzvbK_WHqXroEz5guU/edit?usp=sharing)


Passed | 7
-- | --
Failed | 1
Blocked | 0
Not Executed | 0
Total | 8

1. The failed case is about not writing to `advanced-cache.php` file when hosted in WP Engine.
it looks like we still write to this file during activation. https://i.imgur.com/dVBICJp.png
2. Another issue I noticed is that is not possible to deactivate the plugin when hosted in wordpress.com
any ideas? https://screencast-o-matic.com/watch/cYit3jEBEV

**Comment by Tabrisrp** [comment]
@Davidacu The issue with `advanced-cache.php` should be fixed with my latest commit

For the deactivation issue on wp.com, I looked into it, and it's expected because of the necessity to rename the plugin folder name to bypass the ban. Once the ban is lifted, the deactivation will work correctly.

**Review by Davidacu — CHANGES_REQUESTED** [review]
Looks like I found new issues after this change:

**Pressable**:
1. Notice about writing permissions for `wp-config.php` when trying to deactivate the plugin https://i.imgur.com/47IakuT.png
2. a PHP notice in the debug.log `Undefined property: batcache::$status_header in /wordpress/drop-ins/advanced-cache.php on line 211`

All other cases reported before passed.

**Review by Davidacu — APPROVED** [review]
>Notice about writing permissions for wp-config.php when trying to deactivate the plugin https://i.imgur.com/47IakuT.png

Please disregard the above issue, sorry I was using the wrong build 🤦‍♂️, this is not reproducible in feature branch .

>a PHP notice in the debug.log Undefined property: batcache::$status_header in /wordpress/drop-ins/advanced-cache.php on line 211

And this issue above I discussed with @Tabrisrp is not related to us.
Approving this PR.

**Pull request merged**